### ISSUE TITLE
Documentation of output format and API

### ISSUE THREAD

**Issue by rgrp** [openspending/cubepress#3]
How does this library work and what is the structure of the output files?

**Comment by pudo** [comment]
The idea is to have two input files, a CSV (or other tabular) file and a YAML document that would contain a specification for a) the logical model (i.e. what columns belong together, which ones are measures), b) a database config (if the result isn't supposed to be in-memory) and c) information on which aggregates to generate from the data. The latter is done by specifying a set of filters (e.g. to separate files per year, or for expenditure/revenue), and a set of hierarchies (basically: "break down by cofog1, followed by cofog2, followed by cofog3").

An [example of this spec is here](https://github.com/openspending/cubepress/blob/master/test_data/awards.yaml). Running the script with a skeleton spec file will auto-generate the rest and print it. 

I don't think I ever got around to implementing the output format properly, but it does spit out JSON files which are named after a hashed version of the query used to generate the data. I think that a metadata JSON file and a set of CSV data files might be a better option. 

Finally, I really think that using [babbage](https://github.com/spendb/babbage) would allow this to run basically without any SQL in this package. I'd encourage you to at least have a quick look at the [cube API](https://github.com/spendb/babbage/blob/master/babbage/cube.py#L57). The spec format for the ``cubepress`` and ``babbage``OLAP models is pretty much identical. 

There's also a matured version of the table parsing and loading code [here](https://github.com/spendb/spendb/blob/master/spendb/etl/extract.py).

**Comment by rgrp** [comment]
@pudo really useful. I agree that output as CSV rather than (or in addition to) JSON is probably better.

babbage looks really interesting and, again, spot on in terms of the various options for computing aggregates I've been thinking about:

1. flat-file plus scripting (or even map reduce)
2. load to SQL (OLAP style) and do queries
3. load to Redshift (sort of SQL) and do queries
4. load to BigQuery and do queries
5. ....

As i understand cubepress is basically option 1 and babbage is option 2.

Right now I think one useful thing - and the purpose of the post - was to agree an output format for flat-file aggregates that could be consumed by the viz layer (or views generally) - that was the logic of @danfowler's post https://discuss.okfn.org/t/openspending-next-visualization-experiment/886. This spec could then be common to all of the computation methods. On this I think CSV + (optionally) some kind of JSON descriptor seems a great way to go.

**Comment by pudo** [comment]
Agreed, metadata is everything; the value of having a common spec would be enormous. Without common-format, it will be hard for any visualisation to transfer across aggregates, much less datasets. At the moment, the only visible effect of the lack of metadata is that the treemap shows codes instead of labels, but this would get worse much more quickly if you wanted to add multiple levels, customisable filters etc. 

As for the computing options, I fail to see much value in solving (1) independently of (2): it can easily be expressed as an in-memory SQLite query, which saves you the hassle of having to implement bad, ad-hoc aggregation code. Even (3) can be done as (2) [quite nicely](https://github.com/binarydud/redshift_sqlalchemy).

So what I'm advocating is using cubepress basically as a query generator for babbage or cubes, which freezes the results and generates extra result metadata. The advantage is that you implement the core aggregation logic exactly once, and your models will apply perfectly for both scenarios at the same time.

**Comment by rgrp** [comment]
Agreed re just doing sql on this - it makes a lot of sense.

What is the current output format of cubes or babbage? That's important because that is what visualization would run off and so if we have that specced we can develop viz work and aggregation in tandem with the two kept separate via the API defined in the aggregation output.

**Comment by pudo** [comment]
``babbage`` output is very much oriented after ``cubes``, calling the ``Cube.aggregate`` library function will [return something](https://spendb-dev.herokuapp.com/api/babbage/cubes/md-tenders/aggregate?drilldown=tipul) that is equivalent to a cubes response and vaguely resembles an OpenSpending APIv2 aggregate. It's also very easy to convert the included ``cells`` to CSV.

**Comment by rgrp** [comment]
@pudo can you link the cubes output documentation.

I'm also inlining that output for ref:

```
{
aggregates: [
"_count",
"suma.sum"
],
attributes: [
"tipul.label"
],
page_size: 10000,
summary: {
_count: 138477,
suma.sum: 39809696414.5195
},
order: [ ],
cells: [
{
_count: 45,
suma.sum: null,
tipul.label: null
},
{
_count: 108,
suma.sum: 0,
tipul.label: "Anulat din lipsa finanţării"
},
{
_count: 514,
suma.sum: 98932.65,
tipul.label: "Anulat de AAP"
},
{
_count: 525,
suma.sum: -199201195.31,
tipul.label: "Acord de reziliere"
},
{
_count: 606,
suma.sum: 0,
tipul.label: "Anulat din lipsa trei oferte calificate"
},
{
_count: 836,
suma.sum: 13353490686,
tipul.label: "Anulat din lipsa concurenţei"
},
{
_count: 859,
suma.sum: 0,
tipul.label: "Alt acord"
},
{
_count: 894,
suma.sum: 0,
tipul.label: "Anulat din lipsa ofertelor"
},
{
_count: 1517,
suma.sum: 0,
tipul.label: "Anulat de AC"
},
{
_count: 1853,
suma.sum: 78494.86,
tipul.label: "Acord de prelungirea termenului"
},
{
_count: 2725,
suma.sum: -196189350.01,
tipul.label: "Acord aditional de micsorare/rezeliere"
},
{
_count: 3100,
suma.sum: 462653160.59,
tipul.label: "Acord aditional de majorare"
},
{
_count: 9301,
suma.sum: 240599367.71,
tipul.label: "Acord adiţional"
},
{
_count: 115594,
suma.sum: 26148166318.03,
tipul.label: "Contract de achiziţie"
}
],
cell: [ ],
total_cell_count: 14,
status: "ok",
page: 1
}
```

**Comment by pudo** [comment]
There you go: https://cubes.readthedocs.org/en/latest/server.html#aggregation-and-browsing

**Comment by rgrp** [comment]
@pudo thank-you (i'd poked around but wanted to be sure). Does babbage and cubepress follow this exactly - they look to (which is great).
